### ISSUE TITLE
Passthrough changes Response to be not compliant with WebAssembly.instantiateStreaming

### ISSUE THREAD

**Issue by tschoartschi** [miragejs/miragejs#339]
First of all, we are using Mirage with an Ember project and it works fine 😊thanks for the awesome work. Since we really like Mirage we also tried to get it running for a Vue project which was scaffolded with `vue-cli`. 

Currently, we are struggling to get Cypress running. We use the following plugin `@vue/cli-plugin-e2e-cypress`.  I tried several things but I'm just not able to get our WebAssembly module running.  

We start our WebAssembly module with `WebAssembly.instantiateStreaming` and also if I passthrough the wasm file Chrome complains the following way:

```
Mirage: Passthrough request for GET /static/artifacts/wasm/kernel.wasm
TypeError: Failed to execute 'compile' on 'WebAssembly': An argument must be provided, which must be a Response or Promise<Response> object
```
After some investigation, I think that Mirage wraps every Response in its own Response class? Is this true? For some reason, I wasn't able to figure out why the type-check on `instantiateStreaming` fails. When searching the internet for this error message I only found the following source code: https://doss-gitlab.eidos.ic.i.u-tokyo.ac.jp/kevin/chromium/blob/ad3841a1e47cd90bba10f271a8495d4040490b07/third_party/blink/renderer/bindings/core/v8/v8_wasm_response_extensions.cc#L283

I'm a little bit lost (also because debugging is super hard with the `vue-cli` setup since Cypress is running somewhere else etc).

Any ideas what I could try to get the WebAssembly module running? Basically I followed those two quickstarters:

* https://miragejs.com/quickstarts/vue/development/
* https://miragejs.com/quickstarts/cypress/

Let me know if you need further info

**Comment by tschoartschi** [comment]
I create a repo which should demonstrate the issue, see here: https://github.com/tschoartschi/mirage-passthrough-wasm

The WebAssembly module should print "hello world" to the browsers dev tools console. When Mirage is active but I only define `passthrough` it throws an error. When I do not use Mirage at all everything works fine and "Hello World" is printed to the console as expected.

See the screencast: https://drive.google.com/file/d/1x0mWGyz3uXX2s0an155xn2FsCxAxvERk/view?usp=sharing

Basically I just scaffolded the project with `vue-cli` (https://cli.vuejs.org/guide/creating-a-project.html#vue-create) and added a basic WebAssembly module which I created with emscripten (https://emscripten.org/docs/getting_started/Tutorial.html) 

All the interesting code sits in the files `src/main.js` and `src/App.vue`. Would be great if we could figure out what the problem is 🙂

**Comment by tschoartschi** [comment]
Please just let me quickly know if you are able to reproduce the issue 🙂 if not I could try to provide further information. Thanks a lot

**Comment by samselikoff** [comment]
Sorry about the delayed response! I'm behind on issues a bit due to some consulting/training work.

A repo with a minimal reproduction would definitely help get the issue resolved faster, so if you could provide that it would extremely helpful!

**Comment by tschoartschi** [comment]
@samselikoff no worries 🙂 I created a repo about two weeks ago, see: https://github.com/miragejs/miragejs/issues/339#issuecomment-596625395

Thanks for your great work and let me know if you need anything else

**Comment by samselikoff** [comment]
Copy props https://github.com/pretenderjs/pretender/blob/master/src/index.ts#L126

**Comment by tschoartschi** [comment]
Yesterday I had time to dig into this problem a little more. I think we are facing two issues here:

- the same problem as in (https://github.com/miragejs/ember-cli-mirage/issues/1915). Which basically is a problem that the fake implementation of `fetch` does not correctly send an `arrayBuffer` for requests fetching a WebAssembly file
- the second problem is even trickier: it seems that Chrome recognizes that the `Response` is not a real response and that we send a `FakeResponse`. I think so because if you look at the Chromium source code (https://github.com/chromium/chromium/blob/84525819989a301788f30fe02741c46472ca7772/third_party/blink/renderer/bindings/core/v8/v8_wasm_response_extensions.cc#L312), you can see that the type check fails. The following C++ code does not return a response `Response* response = V8Response::ToImplWithTypeCheck(args.GetIsolate(), args[0]);`. 

It would be great if we could fix those problems but I have no idea 🤔If someone has a hint or some tips/tricks I could have a look at it again but currently I do not know what to do

**Comment by tschoartschi** [comment]
I found a workaround which works at least for us, what I do now is the following:

```JavaScript
new Server({
  environment: 'development',
  logging: true,
  routes() {
    // Needed because Chrome recognizes that the Mirage Response is not a real response
    // with setting instantiateStreaming to null we fallback to legacy WebAssembly instantiation
    // this works with the Mirage Response, therefore the app can start
    // for more details see: https://github.com/miragejs/miragejs/issues/339
    Object.defineProperty(window.WebAssembly, 'instantiateStreaming', {value: null});
    (this.pretender as any).passthroughRequest = (verb: 'GET' | 'POST' | 'PUT' | 'DELETE', path: string, request: any) => {

      // Needed because responseType is not set correctly in Mirages passthrough
      // for more details see: https://github.com/miragejs/ember-cli-mirage/issues/1915
      if (verb === 'GET' && path.match(/\.wasm$/)) {
        console.log('Set responseType for wasm correctly');
        request.responseType = 'arraybuffer';
      }
    };
    this.passthrough();
  },
});

```

@samselikoff maybe you could have a look at this workaround and let me know if this is OK or if it's a stupid idea to do it like this 😉 thanks a lot

**Comment by tschoartschi** [comment]
For everyone who followed this issue, I needed to update the snippet from above slightly. I needed to save the old `passthroughRequest` to a variable so I can call it later inside my extended `passthroughRequest`. See the snippet above and watch out for `const oldPassthroughRequests = (this.pretender as any).passthroughRequest.bind(this.pretender);` 😉

**Comment by tschoartschi** [comment]
@samselikoff since we found workarounds for the problems stated above I tried to follow the Cypress Quick Start guide (https://miragejs.com/quickstarts/cypress/) but now I have again problems with the binary files 🤔 I'm not 💯% what's going on in the quick start for Cypress so I do not know where to apply the patches I described above. 

Maybe you have some time and  could explain in more detail what's happening in the quick start. Maybe this would be also beneficial for others as well 🙂 thanks a lot

**Comment by tschoartschi** [comment]
If someone reads along with this issue I found a workaround which works for us 🙂 basically what we are doing now is the following:

```typescript
import {Server, RestSerializer, Response as MirageResponse} from 'miragejs';
import models from '@/mirage/models';
import fixtures from '@/mirage/data/fixtures';
import {registerRoutes} from '@/mirage/routes';

export function makeServer({environment = 'development', isInsideCrypress = false} = {}) {

  return new Server({

    environment,
    logging: true,

    models,
    fixtures,

    serializers: {
      application: RestSerializer,
    },

    seeds(server) {
      // Load all fixture data into the development db
      server.loadFixtures();
    },

    routes() {

      this.timing = 0; // in dev it defaults to 400 and in test to 0, let's always use 0 to be fast
      this.passthrough();
      this.passthrough('https://url-to-our-cdn/**');

      this.urlPrefix = 'https://url-to-our-api/';
      this.namespace = 'v2/';
      // Needed because Chrome recognizes that the Mirage Response is not a real response
      // with setting instantiateStreaming to null we fallback to legacy WebAssembly instantiation
      // this works with the Mirage Response, therefore the app can start
      // for more details see: https://github.com/miragejs/miragejs/issues/339
      Object.defineProperty(window.WebAssembly, 'instantiateStreaming', {value: null});
      const oldPassthroughRequests = (this.pretender as any).passthroughRequest.bind(this.pretender);
      (this.pretender as any).passthroughRequest = (verb: 'GET' | 'POST' | 'PUT' | 'DELETE', path: string, request: any) => {
        // Needed because responseType is not set correctly in Mirages passthrough
        // for more details see: https://github.com/miragejs/miragejs/issues/1915
        if (verb === 'GET' && path.match(/\.wasm$/)) {
          request.responseType = 'arraybuffer';
        }
        return oldPassthroughRequests(verb, path, request);
      };
      if (isInsideCrypress) {
        const methods = ['get', 'put', 'patch', 'post', 'delete'];
        methods.forEach((method) => {
          (this as any)[method]('/*', async (_schema: any, request: any) => {
            const [status, headers, body] = await (window as any).handleFromCypress(request);
            return new MirageResponse(status, headers, body);
          });
        });
      } else {
        registerRoutes(this as any);
      }

    },
  });

}
```

And then in `main.ts` we do the following:

```typescript
import {makeServer} from '@/mirage/server';
if ((window as any).Cypress) {
  makeServer({isInsideCrypress: true});
} else {
  makeServer();
}
```

and in the specs we do the following:

```typescript
import {makeServer} from '@/mirage/server';
let server;

beforeEach(() => {
  cy.log('Setup mirage server');
  server = makeServer({environment: 'test'});
});

afterEach(() => {
  cy.log('Shutdown mirage server');
  server.shutdown();
});
```
Now everything is running 🙂 

As a side-note I wanted to share my other tries: at first, I tried to "monkey patch" the serialize function in `server.js` https://github.com/miragejs/miragejs/blob/80bc81a3d466ef88a0da63f9090f650119fd56fd/lib/server.js#L1037 because since we forward everything to `handleFromCypress` with `this[method]('/*', ...)` also our requests to our `WebAssembly` files were `JSON.stringified`. But then I run into the next problem. Because the npm package `fake xml http request` throws an exception if a body is not a string. This is done by `verifyResponseBodyType`. I'm not an expert with `XMLHttpRequest` and therefore I do not know if it's a must that we convert everything to string before sending a `XMLHttpRequest`. At least in the case of a `WebAssembly` file this doesn't seem to be beneficial. 

Now for us, it works because we do not need to "mock" other things than JSON requests but maybe this could become a problem for other people.

**Comment by ryanto** [comment]
Thanks for the detailed explanation @tschoartschi ! 

It seems that setting the request's responseType, and then allowing the passthrough to happen is what's needed to get your code running as expected. Is that correct?

How your are making the initial request in your application. I'd like to see the  fetch call that's request to wasm file.

**Comment by tschoartschi** [comment]
@ryanto thanks for the response and sorry for the late reply but yesterday was a national public holiday in Austria. I created a repo which demonstrates the issue: https://github.com/miragejs/miragejs/issues/339#issuecomment-596625395

Let me know if this is enough our if you need further information 🙂 

I think it's a bug 🐛 in FakeXMLHttpRequest because I think FakeXMLHttpRequest does not set the `responseType` correctly. When debugging I found no sane way to tell FakeXMLHttpRequest to set the responseType. The only way I found was to set the rsponseType based on the file extension. This is what I outlined above in the `passthrough` hack

**Comment by raldone01** [comment]
🔥 @tschoartschi 🔥 Thank you! You are amazing! I hope this issue gets fixed soon. It's critical for my application.

**Comment by mbrevda** [comment]
@samselikoff any progress on this? It's currently a blocker for us and we've invested so much with Mirage, would be a huge shame to have to rip it all out...

**Comment by samselikoff** [comment]
@mbrevda unfortunately no, I have not worked on this issue at all.

If you do end up diving into FakeXMLHttpRequest I can help you get any PRs merged!

**Comment by mbrevda** [comment]
The issue is actually with passthrough request, not those served by mirage. Would FakeXMLHttpRequest be the right place to look?

**Comment by tschoartschi** [comment]
@mbrevda I really deep-dived on this issue because it was also a blocker for us. I was able to work around the problem. I explained how I did it here: https://github.com/miragejs/miragejs/issues/339#issuecomment-626619513

but the gist is:

```typescript
// Needed because Chrome recognizes that the Mirage Response is not a real response
// with setting instantiateStreaming to null we fallback to legacy WebAssembly instantiation
// this works with the Mirage Response, therefore the app can start
// for more details see: https://github.com/miragejs/miragejs/issues/339
Object.defineProperty(window.WebAssembly, 'instantiateStreaming', {value: null});
const oldPassthroughRequests = (this.pretender as any).passthroughRequest.bind(this.pretender);
(this.pretender as any).passthroughRequest = (verb: 'GET' | 'POST' | 'PUT' | 'DELETE', path: string, request: any) => {
  // Needed because responseType is not set correctly in Mirages passthrough
  // for more details see: https://github.com/miragejs/miragejs/issues/1915
  if (verb === 'GET' && path.match(/\.wasm$/)) {
    request.responseType = 'arraybuffer';
  }
  return oldPassthroughRequests(verb, path, request);
};
```

And as I wrote above I think it's a problem with FakeXMLHttpRequest. I'm also not sure if there is any chance to get it running with `instantiateStreaming` this is why I did: `Object.defineProperty(window.WebAssembly, 'instantiateStreaming', {value: null})`.

Would be great to see if someone could fix this bug 🙂 currently, I do not have the time to work on it (also because the hack works for us "good enough" at the moment).

**Comment by mbrevda** [comment]
Monkeypatched as above it works. Would be good to have proper support in place!
